apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs

```diff
@@ -0,0 +1,430 @@
+    use proptest::prelude::*;
+
+    use super::states::{NextState, NextStateWithEncryptedTxs, TryAlloc};
+    use super::*;
+    use crate::node::ledger::shims::abcipp_shim_types::shim::TxBytes;
+
+    /// Proptest generated txs.
+    #[derive(Debug)]
+    struct PropTx {
+        tendermint_max_block_space_in_bytes: u64,
+        protocol_txs: Vec<TxBytes>,
+        encrypted_txs: Vec<TxBytes>,
+        decrypted_txs: Vec<TxBytes>,
+    }
+
+    proptest! {
+        /// Check if we reject a tx when its respective bin
+        /// capacity has been reached on a [`BlockSpaceAllocator`].
+        #[test]
+        fn test_reject_tx_on_bin_cap_reached(max in prop::num::u64::ANY) {
+            proptest_reject_tx_on_bin_cap_reached(max)
+        }
+
+        /// Check if the sum of all individual bin allotments for a
+        /// [`BlockSpaceAllocator`] corresponds to the total space ceded
+        /// by Tendermint.
+        #[test]
+        fn test_bin_capacity_eq_provided_space(max in prop::num::u64::ANY) {
+            proptest_bin_capacity_eq_provided_space(max)
+        }
+
+        /// Test that dumping txs whose total combined size
+        /// is less than the bin cap does not fill up the bin.
+        #[test]
+        fn test_tx_dump_doesnt_fill_up_bin(args in arb_transactions()) {
+            proptest_tx_dump_doesnt_fill_up_bin(args)
+        }
+    }
+
+    /// Implementation of [`test_reject_tx_on_bin_cap_reached`].
+    fn proptest_reject_tx_on_bin_cap_reached(
+        tendermint_max_block_space_in_bytes: u64,
+    ) {
+        let mut bins =
+            BlockSpaceAllocator::init(tendermint_max_block_space_in_bytes);
+
+        // fill the entire bin of decrypted txs
+        bins.decrypted_txs.occupied_space_in_bytes =
+            bins.decrypted_txs.allotted_space_in_bytes;
+
+        // make sure we can't dump any new decrypted txs in the bin
+        assert_matches!(
+            bins.try_alloc(b"arbitrary tx bytes"),
+            AllocStatus::Rejected { .. }
+        );
+    }
+
+    /// Implementation of [`test_bin_capacity_eq_provided_space`].
+    fn proptest_bin_capacity_eq_provided_space(
+        tendermint_max_block_space_in_bytes: u64,
+    ) {
+        let bins =
+            BlockSpaceAllocator::init(tendermint_max_block_space_in_bytes);
+        assert_eq!(0, bins.uninitialized_space_in_bytes());
+    }
+
+    /// Implementation of [`test_tx_dump_doesnt_fill_up_bin`].
+    fn proptest_tx_dump_doesnt_fill_up_bin(args: PropTx) {
+        let PropTx {
+            tendermint_max_block_space_in_bytes,
+            protocol_txs,
+            encrypted_txs,
+            decrypted_txs,
+        } = args;
+
+        // produce new txs until the moment we would have
+        // filled up the bins.
+        //
+        // iterate over the produced txs to make sure we can keep
+        // dumping new txs without filling up the bins
+
+        let bins = RefCell::new(BlockSpaceAllocator::init(
```

Comment:
Why is the RefCell needed?

Reply:
the iterator borrows the allocator immutably, then while iterating we borrow it mutably. without the `RefCell`, we'd get the borrow checker yelling at us